Show optional parameters with checkbox

emmarogge opened the pull request:
Show optional parameters only when the widget's checkbox has been selected. Also refactor widget code for brevity.

emmarogge committed:
Show optional parameters with checkbox

Show optional parameters only when the widget's checkbox has been selected. Also refactor widget code for brevity.

wnojopra commented on first_hour_on_vwb/working_with_resources.ipynb:
```diff
@@ -855,58 +855,38 @@
+    "        self.toggle_optional = widgets.Checkbox(False, description = \"Show optional inputs\",indent=False)\n",
+    "        self.toggle_optional.observe(self.toggle)\n",
+    "        self.vb = widgets.VBox(\n",
+    "            children = [self.toggle_optional,self.label,self.warning,self.access_drop_down,self.cloning_drop_down,self.input_description,self.input_name,self.button,self.output], \n",
```
Long lists on one line aren't ideal because 
1) I need to scroll horizontally to see the whole list
2) It's more error prone when editing in the future due to having to find a particular spot in one line to edit. 

I would highly recommend having it back to the one item per line style. Also acceptable would be having a few items per line, like
```
children = [
  self.toggle_optional, self.label, self.warning,
  self.access_drop_down, self.cloning_drop_down, ...etc
]

Also there's currently no whitespace after commas, which really goes against readability. 

Repeat this comment for the other related sections.

emmarogge replied:
fixed!

wnojopra commented on first_hour_on_vwb/working_with_resources.ipynb:
```diff
@@ -855,58 +855,38 @@
     "\n",
+    "    def toggle(self,event):\n",
+    "        if self.toggle_optional.value == True:\n",
+    "            self.vb.children = [self.toggle_optional,self.label,self.warning,self.access_drop_down,self.cloning_drop_down,self.input_dataset_id,self.input_description,self.input_name,self.button,self.output]\n",
```
The long lists in this class are _mostly_ repeated. Could we declare a long list of common children once, and then modify as needed? Something like...

```
self.common_children = [self.toggle_optional, self.label, self.warning, self.access_drop_down ...]

self.vb.children = self.common_children.insert(6, self.input_dataset_id)
```

Repeat this comment for the other related sections.

emmarogge replied:
fixed!

wnojopra commented on first_hour_on_vwb/working_with_resources.ipynb:
```diff
@@ -855,58 +855,38 @@
+    "        self.toggle_optional.observe(self.toggle)\n",
+    "        self.vb = widgets.VBox(\n",
+    "            children = [self.toggle_optional,self.label,self.warning,self.access_drop_down,self.cloning_drop_down,self.input_description,self.input_name,self.button,self.output], \n",
+    "            layout=wu.vbox_layout\n",
```
Nit: Lets be consistent with whitespacing around the equal sign

emmarogge replied:
fixed!

wnojopra reviewed (changes requested):
This is awesome! Do you know if its preferable to have it be "Show optional inputs" vs "Hide optional inputs"? I personally don't care either way but I wonder if there has been some UX research done on this. 

Otherwise just a few styling changes to make inline

emmarogge committed:
Improve readability of lists of widget inputs

To improve readability, use consistent spacing and refactor to make entire list of widgets visible without horizontal scrolling. Additionally, refactor widgets to use a single list instead of creating additional copies of the list.

emmarogge commented:
>Do you know if its preferable to have it be "Show optional inputs" vs "Hide optional inputs"?
> 

I once read an explanation that showing optional fields by default (vs. allowing users to choose to see them) is essentially asking users to deal with unnecessary information. I think that's true in this case, so I'd prefer not to show the optional fields in the initial widget state. If we get feedback about this, we can definitely decide to pivot.

wnojopra commented on first_hour_on_vwb/widget_utils.py:
```diff
@@ -101,4 +101,16 @@ def __post_init__(self):
+    def __init__(self):
+        self.checkbox = widgets.Checkbox(
+            False, 
+            description = "Show optional parameters",
```
While we're here, lets fixup the spacing around equal signs in this file to be consistent.

emmarogge replied:
I have installed and run the Google Python linter/formatter on this file now, so that it's self-consistent.

wnojopra commented on first_hour_on_vwb/working_with_resources.ipynb:
```diff
@@ -855,58 +855,49 @@
+    "                parameterList = []\n",
+    "                for r in self.required_fields:\n",
+    "                    parameterList.append(r)\n",
+    "                parameterList.insert(5,self.input_dataset_id)\n",
```
The 5 here is a magic number. We at should at least have a comment explaining it. Something like "Inserting parameters at specific locations in the UI".

emmarogge replied:
Addressed!

emmarogge committed:
Clean up Python file and address magic numbers

Used the Google formatting/linting tool to ensure all spacing is
consistent in widget utilities Python module. Commented the reasoning
for magic numbers in widgets with optional parameters.
